Add mode to MemoryDep to track atomic accumulates

isuruf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123222
* __->__ #123223

And allow fusion of buffers where writes are only atomic accumulates.
This allows fusing of ops like

  _unsafe_index_put(_unsafe_index_put(a, ...), ...)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

isuruf committed:
Introduce an AccumulateDep to track atomic accumulates

And allow fusion of buffers where writes are only atomic accumulates.
This allows fusing of ops like

  _unsafe_index_put(_unsafe_index_put(a, ...), ...)

[ghstack-poisoned]

isuruf committed:
Update on "Introduce an AccumulateDep to track atomic accumulates"


And allow fusion of buffers where writes are only atomic accumulates.
This allows fusing of ops like

  _unsafe_index_put(_unsafe_index_put(a, ...), ...)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

isuruf commented:
cc @peterbell10 @lezcano

isuruf commented:
The failing test is real. It's a dependency tracking issue. Looking at it now.

peterbell10 commented on torch/_inductor/dependencies.py:
```diff
@@ -280,7 +342,7 @@ def load_seed(self, name: str, index: int):
 
     def store(self, name: str, index: sympy.Expr, value: str, mode=None) -> str:
-        self._writes.add(MemoryDep(name, *self.canonicalize(index)))
+        self._writes_with_mode[mode].add(MemoryDep(name, *self.canonicalize(index)))
```
Would it make more sense to represent it this way?
```python
if mode is None:
    dep = MemoryDep(name, *self.canonicalize(index))
    self._writes.add(dep)
else:
    dep = AccumulateDep(name, *self.canonicalize(index), mode)
    self._reads.add(dep)
    self._writes.add(dep)
```

`atomic_add` is a "read-modify-write" op so it makes sense to me that it would add a read and a write dependency. Making it a different class should also prevent accidentally treating it as the same as a normal read/write.

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -2060,6 +2066,8 @@ def can_fuse_vertical(self, node1, node2):
         # we still can match unmet dep
         # if there's indirect indexing, don't match it
         def fusable_read_and_write(read: Dep, write: Dep):
```
Side note: tracking `AccumulateDep` differently from `MemoryDep` should allow us to remove this hacky fix that blocks `atomic_add` from participating in vertical fusion 

https://github.com/pytorch/pytorch/blob/0c8a165b43c5cc226a84be46b8049e5dfca18aca/torch/_inductor/scheduler.py#L2073-L2090

isuruf committed:
Update on "Introduce an AccumulateDep to track atomic accumulates"


And allow fusion of buffers where writes are only atomic accumulates.
This allows fusing of ops like

  _unsafe_index_put(_unsafe_index_put(a, ...), ...)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -2106,25 +2121,6 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
-            if any(
-                (
-                    node2_used_buf in self.mutation_renames
-                    and node1.has_atomic_add(self.mutation_renames[node2_used_buf])
```
You should be able to remove `has_atomic_add` as well.

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -1691,6 +1704,8 @@ def compute_ancestors(self):
             ancestors = set()
             for dep in node.unmet_dependencies:
+                if dep.name == node.get_name():
+                    continue
```
It seems to be the case that we're just ignoring the read `MemoryDep` everywhere that it comes up. Do the tests pass if we remove it entirely and just rely on the `StarDep`?

isuruf replied:
Where do you suggest we remove the MemoryDep?

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -2189,6 +2185,23 @@ def can_fuse_vertical(self, node1, node2):
+                and isinstance(write, MemoryDep)
+                and read.mode == write.mode
+                and write.mode is not None
+                and read_name == write_name
```
Do we have any tests where we do something like:
```python
x.index_add_(...)
y = x.clone()
x.index_add_(...)
return x.clone(), y
```

and see that the two `index_add_` operations don't get fused into the same kernel?

isuruf committed:
Update

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/dependencies.py:
```diff
@@ -338,7 +344,9 @@ def load_seed(self, name: str, index: int):
-        self._writes.add(MemoryDep(name, *self.canonicalize(index)))
+        self._writes.add(MemoryDep(name, *self.canonicalize(index), mode=mode))
+        if mode is not None:
+            self._reads.add(MemoryDep(name, *self.canonicalize(index), mode=mode))
```
This is the dep I think we might be able to remove.

isuruf replied:
Tests pass with this dep removed. Thanks. Ready for a review

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -1486,6 +1474,7 @@ def dep_closure(node_name):
             for read_dep in node.read_writes.reads:
                 if (
                     read_dep.name in self.name_to_node
+                    and read_dep.name != node_name
```
```suggestion
```

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -1535,18 +1524,30 @@ def add_user(used_by_name, user_node, can_inplace=False, is_weak=False):
                     other_name = rename(other_node.get_name())
                     known_dep_node_names = dep_closure(node.get_name())
-                    if other_name not in known_dep_node_names:
+                    if (
+                        other_name != alt_name
+                        and other_name not in known_dep_node_names
+                    ):
```
```suggestion
                    if other_name not in known_dep_node_names:
```

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -1703,6 +1704,8 @@ def compute_ancestors(self):
         for node in self.nodes:
             ancestors = set()
             for dep in node.unmet_dependencies:
+                if dep.name == node.get_name():
+                    continue
```
```suggestion
```

peterbell10 reviewed (commented):
I believe these are no longer needed.

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
@pytorchbot merge
